packages/js/src/client.ts

```diff
@@ -95,33 +95,75 @@ class BaseClient extends HTTPClient {
    *
    * @param apl - the apl query
    * @param options - optional query options
-   * @returns result of the query, check: {@link QueryResult}
+   * @returns result of the query depending on the format in options, check: {@link QueryResult} and {@link TabularQueryResult}
    *
    * @example
    * ```
    * await axiom.query("['dataset'] | count");
    * ```
    *
    */
-  query = (apl: string, options?: QueryOptions): Promise<QueryResult> => {
+  query = <
+    TOptions extends QueryOptions,
+    TResult = TOptions['format'] extends 'tabular' ? Promise<TabularQueryResult> : Promise<QueryResult>,
+  >(
+    apl: string,
+    options?: TOptions,
+  ): Promise<TResult> => {
     const req: Query = { apl: apl };
     if (options?.startTime) {
       req.startTime = options?.startTime;
     }
     if (options?.endTime) {
       req.endTime = options?.endTime;
     }
-    return this.client.post<QueryResult>(
-      this.localPath + '/datasets/_apl',
-      {
-        body: JSON.stringify(req),
-      },
-      {
-        'streaming-duration': options?.streamingDuration as string,
-        nocache: options?.noCache as boolean,
-        format: 'legacy',
-      },
-    );
+
+    return this.client
+      .post<TOptions['format'] extends 'tabular' ? RawTabularQueryResult : QueryResult>(
+        this.localPath + '/datasets/_apl',
+        {
+          body: JSON.stringify(req),
+        },
+        {
+          'streaming-duration': options?.streamingDuration as string,
+          nocache: options?.noCache as boolean,
+          format: options?.format ?? 'legacy',
+        },
+      )
+      .then((res) => {
+        if (options?.format !== 'tabular') {
+          return res;
+        }
+
+        const result = res as RawTabularQueryResult;
+        return {
+          ...res,
+          tables: result.tables.map((t) => {
+            return {
+              ...t,
+              events: function* () {
```

Comment:
> lgtm, but would be nice to have some way of easily getting to the events. the other sdks expose an iterator for each table, could do something similar.

@bahlo Added events iterator. I am not sure if this is the best implementation, and one thing that's on my mind is this being a pain when doing SSR/Serialization, happy to discuss it further

Reply:
Huh, good catch wrt SSR/Serialization. We could make tables a class that implements `toJSON` maybe?